x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/NativeAnalyticsProcess.java

```diff
@@ -56,10 +59,23 @@ public AnalyticsProcessConfig getConfig() {
     }
 
     @Override
-    public void restoreState(BytesReference state) throws IOException {
-        Objects.requireNonNull(state);
+    public void restoreState(Client client, String stateDocIdPrefix) throws IOException {
+        Objects.requireNonNull(stateDocIdPrefix);
         try (OutputStream restoreStream = processRestoreStream()) {
-            StateToProcessWriterHelper.writeStateToStream(state, restoreStream);
+            int docNum = 0;
+            while (true) {
+                if (isProcessKilled()) {
+                    return;
+                }
+
+                SearchResponse stateResponse = client.prepareSearch(AnomalyDetectorsIndex.jobStateIndexPattern())
+                    .setSize(1)
+                    .setQuery(QueryBuilders.idsQuery().addIds(stateDocIdPrefix + ++docNum)).get();
```

Comment:
This comment is somewhat related to the one about using recursion. Could we fetch all the state docs at once using some pattern-matched query?
If not, and if 
we expect  many state docs to exist, we could try to fetch them in batches (by adding ids `#1`, `#2`, ..., `#10` to one ids query) to reduce the number of queries.
Please disregard the above if we expect the number of docs to be small so there is no need for premature optimization.

Reply:
> Could we fetch all the state docs at once using some pattern-matched query?

We deliberately fetch one at a time because if there is more than one then all but the last will be quite big (~16MB in size).  The reason why we split the state over multiple documents is to avoid using all the JVM heap with a massive model state that we have to hold in memory in its entirety.

There should probably be a comment to say we are deliberately only bringing one document into memory at a time because they are big.

Reply:
> The reason why we split the state over multiple documents is to avoid using all the JVM heap with a massive model state that we have to hold in memory in its entirety.

Great, thanks for explaining that!